config/ax_pthread.m4

```diff
@@ -0,0 +1,486 @@
+#
+#   If you are only building threaded programs, you may wish to use these
+#   variables in your default LIBS, CFLAGS, and CC:
+#
+#     LIBS="$PTHREAD_LIBS $LIBS"
+#     CFLAGS="$CFLAGS $PTHREAD_CFLAGS"
+#     CC="$PTHREAD_CC"
+#
+#   In addition, if the PTHREAD_CREATE_JOINABLE thread-attribute constant
+#   has a nonstandard name, this macro defines PTHREAD_CREATE_JOINABLE to
+#   that name (e.g. PTHREAD_CREATE_UNDETACHED on AIX).
+#
+#   Also HAVE_PTHREAD_PRIO_INHERIT is defined if pthread is found and the
+#   PTHREAD_PRIO_INHERIT symbol is defined when compiling with
+#   PTHREAD_CFLAGS.
+#
+#   ACTION-IF-FOUND is a list of shell commands to run if a threads library
+#   is found, and ACTION-IF-NOT-FOUND is a list of commands to run it if it
+#   is not found. If ACTION-IF-FOUND is not specified, the default action
+#   will define HAVE_PTHREAD.
+#
+#   Please let the authors know if this macro fails on any platform, or if
+#   you have any other suggestions or comments. This macro was based on work
+#   by SGJ on autoconf scripts for FFTW (http://www.fftw.org/) (with help
+#   from M. Frigo), as well as ac_pthread and hb_pthread macros posted by
+#   Alejandro Forero Cuervo to the autoconf macro repository. We are also
+#   grateful for the helpful feedback of numerous users.
+#
+#   Updated for Autoconf 2.68 by Daniel Richard G.
+#
+# LICENSE
+#
+#   Copyright (c) 2008 Steven G. Johnson <stevenj@alum.mit.edu>
+#   Copyright (c) 2011 Daniel Richard G. <skunk@iSKUNK.ORG>
+#
+#   This program is free software: you can redistribute it and/or modify it
+#   under the terms of the GNU General Public License as published by the
+#   Free Software Foundation, either version 3 of the License, or (at your
+#   option) any later version.
+#
+#   This program is distributed in the hope that it will be useful, but
+#   WITHOUT ANY WARRANTY; without even the implied warranty of
+#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
+#   Public License for more details.
+#
+#   You should have received a copy of the GNU General Public License along
+#   with this program. If not, see <https://www.gnu.org/licenses/>.
+#
+#   As a special exception, the respective Autoconf Macro's copyright owner
+#   gives unlimited permission to copy, distribute and modify the configure
+#   scripts that are the output of Autoconf when processing the Macro. You
+#   need not follow the terms of the GNU General Public License when using
+#   or distributing such scripts, even though portions of the text of the
+#   Macro appear in them. The GNU General Public License (GPL) does govern
+#   all other use of the material that constitutes the Autoconf Macro.
+#
+#   This special exception to the GPL applies to versions of the Autoconf
+#   Macro released by the Autoconf Archive. When you make and distribute a
+#   modified version of the Autoconf Macro, you may extend this special
+#   exception to the GPL to apply to your modified version as well.
```

Comment:
This file might be a deal-breaker for hwloc.

hwloc is distributed under the BSD license.  This file specifically states that it is GPL v3 or later.  GPL v3 is... difficult to work with.

There are the "special exceptions", but that seems to only cover "*the configure scripts that are the output of Autoconf when processing the Macro*."  To me (and I am *not* a lawyer / this is not legal advice / yadda yadda yadda), this sounds like we would then have a GPL v3 file in our repo (ouch).  I don't recall offhand if hwloc includes its `.m4` files in distribution tarballs (I think it does?), which would mean that there is a GPL v3 file in an otherwise-BSD tarball.  That gets tricky to explain to people.

I don't know if the above analysis is correct or not, but why even introduce the question?  Why raise the spectre of GPL v3?

My $0.02: just write the tests yourself and stay away from the AX stuff.

Reply:
This question holds for all autoconf bits, it's not just ax_pthread.m4. See for instance config/libtool.m4, it's the same kind of text.

Put another way, we already do distribute gpl files in the hwloc tarballs since long :)

Reply:
No, it's still different: libtool.m4 is not in our repo.

Reply:
But it's distributed in our tarballs.
Actually, we could even not put ax_pthread.m4 in the repo, and assume that people running autogen.sh have autoconf-archive available.

Reply:
I'm admittedly taking a fine line here between in the repo / generated at autogen.sh time...

But GPL3 is a whole spectre in itself.  Suffice it to say that GPL3 will... be difficult.  I don't know what the restrictions are, but ... I don't even want to find out.  😄 

I also don't think it's a good assumption that people will have autoconf-archive already available (e.g., I don't).  It would be another requirement.

There's also downstream considerations -- Open MPI, MPICH, etc.  Anyone who embeds hwloc.

Let me take a step back and ask: what exactly is the need for `AX_PTHREAD`?  Can't it be re-written in a simple test without raising any of these questions?

Reply:
But I mean the question is already raised by the presence of libool.m4, which is definitely GPLv2 with the same exception as ax_pthread.m4

Reply:
Let me take a step back and ask: what exactly is the need for `AX_PTHREAD`? Can't it be re-written in a simple test without raising any of these questions?

Reply:
AX_PTHREAD is for detection of pthread capabilities and defining macros to setup CFLAGS and LDFLAGS for compilation. In my opinion it is not a big deal to remove AX_PTHREAD and 
make a simple replacement instead. I assumed that all the boring portability checks would have been thoroughly performed there and I would not have reinvent the will. I didn't check the license though.. I am going to replace it with something more convenient licence-wise.

Reply:
I updated with the original check from hwloc master.
See #https://github.com/open-mpi/hwloc/pull/359/commits/a7979a9c2aa3fd277a38c239432dcfb8db97202b

Reply:
Please hit resolve if it suits you.